### ISSUE TITLE
Truststore, keystore, ldaps and CLI tool

### ISSUE THREAD

**Issue by richiereynolds** [rundeck/rundeck#2099]
# Bug report

**My Rundeck detail**
- Rundeck version: 2.6.9
- install type: rpm (rpm,deb,war,launcher)
- OS Name/version: RH6.7

**Expected Behavior**
Ability to use ldaps, ssl and cli tools together

---

**Actual Behavior**
Pick 2 out of 3

---

**How to reproduce Behavior**
Attempt to use all 3

I create an /etc/rundeck/ssl/truststore file using my company's cert. which allow me to use ldaps, following http://rundeck.org/docs/administration/authenticating-users.html

I generate a /etc/rundeck/ssl/keystore file to use ssl following http://rundeck.org/docs/administration/configuring-ssl.html

This works fine. But now I want to use CLI tool, the guide says:

CLI tools that communicate to the Rundeck server need to trust the SSL certificate provided by the server. They are preconfigured to look for a truststore at the location: $RDECK_BASE/etc/truststore. Copy the keystore as the truststore for CLI tools:

cp etc/keystore etc/truststore

For RPM installs, I take this to mean:
cp /etc/rundeck/ssl/keystore  /etc/rundeck/ssl/truststore 
which wipes out the truststore I've just created for ldaps!

Am I just being stupid?

**Comment by gschueler** [comment]
you should be able to import both certs to the truststore instead.  the copy step is a shortcut if you are simply relying on a local ssl key, and because the CLI tools are preconfigured to use that truststore location.  You could use a different location as well

**Comment by richiereynolds** [comment]
By import, do you mean running:
keytool -keystore /etc/rundeck/ssl/truststore -alias rundeck -genkey -keyalg RSA -keypass adminadmin -storepass adminadmin
again, and entering the same vals as I did when generating /etc/rundeck/ssl/keystore?

If so, i tried that and restarted RD. I can still do ldaps and https login but I still can't run the cli tool.
Error: Error making server request to https://localhost:4440: Error occurred while trying to authenticate to server: sun.security.validator.ValidatorException: PKIX path validation failed: java.security.cert.CertPathValidatorException: signature check failed

(Note that I've set HTTPS_PORT to 4440)

When you say I could use a different location, I take it you mean e.g using trustore.ldap to store my ldaps cert.
I tried this too but still couldn't get all functions to work at once. Which settings do i need to change to do this? I've found:
/etc/rundeck/ssl/ssl.properties: has keystore and truststore
/etc/rundeck/profile: has -Djavax.net.ssl.trustStore

What's used by which application?

**Comment by richiereynolds** [comment]
Got it to work via the import method.

First, i had to take the keystore I had generated for ssl and generate an X509 certificate from it:
keytool -export -alias rundeck -keystore keystore -rfc -file self_signed_X509.cer

Then I imported that cert into the truststore where my ldap key is:
keytool -import -alias rundeck -file self_signed_X509.cer -keystore /etc/rundeck/ssl/truststore

Now all 3 functions work. I think the documentation around this area is a little lacking e.g. I found  had to add export RDECK_JVM="$RDECK_JVM $RDECK_SSL_OPTS" to profile on some forum. Where the launcher and RH installations differ is also slightly unclear in places.

The hardcoding of the location of the truststore for the cli location also seems a bit limiting.
